presto-main/src/main/java/io/prestosql/operator/scalar/ArraysOverlapFunction.java

```diff
@@ -21,80 +21,93 @@
 
 import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
+import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
 import static io.prestosql.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
 import static io.prestosql.spi.function.OperatorType.COMPARISON;
-import static io.prestosql.spi.type.BigintType.BIGINT;
-import static io.prestosql.spi.type.IntegerType.INTEGER;
 
 @ScalarFunction("arrays_overlap")
 @Description("Returns true if arrays have common elements")
 public final class ArraysOverlapFunction
 {
-    private int[] leftPositions;
-    private int[] rightPositions;
+    private int[] leftPositions = new int[0];
+    private int[] rightPositions = new int[0];
 
-    private long[] leftLongArray;
-    private long[] rightLongArray;
+    private long[] leftLongArray = new long[0];
+    private long[] rightLongArray = new long[0];
 
     @TypeParameter("E")
     public ArraysOverlapFunction(@TypeParameter("E") Type elementType) {}
 
-    private static IntComparator intBlockCompare(BlockPositionComparison comparisonOperator, Block block)
-    {
-        return new AbstractIntComparator()
-        {
-            @Override
-            public int compare(int left, int right)
-            {
-                if (block.isNull(left) && block.isNull(right)) {
-                    return 0;
-                }
-                if (block.isNull(left)) {
-                    return 1;
-                }
-                if (block.isNull(right)) {
-                    return -1;
-                }
-                return (int) comparisonOperator.compare(block, left, block, right);
-            }
-        };
-    }
-
     @SqlNullable
+    @TypeParameter("E")
+    @TypeParameterSpecialization(name = "E", nativeContainerType = long.class)
     @SqlType(StandardTypes.BOOLEAN)
     public Boolean arraysOverlapInt(
             @OperatorDependency(
                     operator = COMPARISON,
-                    argumentTypes = {"integer", "integer"},
-                    convention = @Convention(arguments = {BLOCK_POSITION, BLOCK_POSITION}, result = FAIL_ON_NULL)) BlockPositionComparison comparisonOperator,
-            @SqlType("array(integer)") Block leftArray,
-            @SqlType("array(integer)") Block rightArray)
+                    argumentTypes = {"E", "E"},
+                    convention = @Convention(arguments = {NEVER_NULL, NEVER_NULL}, result = FAIL_ON_NULL)) LongComparison comparisonOperator,
+            @TypeParameter("E") Type type,
+            @SqlType("array(E)") Block leftArray,
+            @SqlType("array(E)") Block rightArray)
     {
-        return genericArraysOverlap(leftArray, rightArray, INTEGER);
+        int leftSize = leftArray.getPositionCount();
+        int rightSize = rightArray.getPositionCount();
+
+        if (leftSize == 0 || rightSize == 0) {
+            return false;
```

Comment:
This seems like a (pre-existing) bug. If two arrays are equal to each other, then `arrays_overlap` should return true. Therefore, since `array[] == array[]` => `arrays_overlap(array[], array[])` should return `true`.

It's especially confusing that `arrays_overlap(x, x)` might ever return `false`.